### PR TITLE
net/tstun: add inital support for NAT v4

### PR DESCRIPTION
**Pull request by maisem**
This adds support in tstun to utitilize the SelfNodeV4MasqAddrForThisPeer and
perform the necessary modifications to the packet as it passes through tstun.

Currently this only handles ICMP, UDP and TCP traffic.
Subnet routers and Exit Nodes are also unsupported.

Updates tailscale/corp#8020

Co-authored-by: Melanie Warrick <warrick@tailscale.com>
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: net/packet/packet.go [hunk 1]
```diff
@@ -440,6 +440,40 @@ func (q *Parsed) IsEchoResponse() bool {
 	}
 }
 
+// UpdateSrcAddr updates the source address in the packet buffer (e.g. during
+// SNAT). It also updates the checksum. Currently (2022-12-10) only TCP/UDP/ICMP
+// over IPv4 is supported. It panics if called with IPv6 addr.
+func (q *Parsed) UpdateSrcAddr(src netip.Addr) {
+	if q.IPVersion != 4 || src.Is6() {
+		panic("UpdateSrcAddr: only IPv4 is supported")
```
Comment (twitchyliquid64) [inline net/packet/packet.go:448]: This is a panic but the equivalent guard in `UpdateDstAddr` is a silent return - be consistent (probably a panic?)

File: net/packet/packet.go [hunk 2]
```diff
@@ -502,3 +536,53 @@ func withIP(ap netip.AddrPort, ip netip.Addr) netip.AddrPort {
 func withPort(ap netip.AddrPort, port uint16) netip.AddrPort {
 	return netip.AddrPortFrom(ap.Addr(), port)
 }
+
+// updateV4PacketChecksums updates the checksums in the packet buffer.
+// Currently (2023-03-01) only TCP/UDP/ICMP over IPv4 is supported.
+// The packet is modified in place.
+func updateV4PacketChecksums(p *Parsed, old, new netip.Addr) {
```
Comment (twitchyliquid64) [inline net/packet/packet.go:543]: Add to docs that the `*Parsed` must have been decoded as an IPProto != Unknown or this is unsafe.

File: net/packet/packet.go [hunk 3]
```diff
@@ -502,3 +536,53 @@ func withIP(ap netip.AddrPort, ip netip.Addr) netip.AddrPort {
...
+		updateV4Checksum(p.Transport()[16:18], o4[:], n4[:])
+	case ipproto.ICMPv4:
+		// Nothing to do.
+	}
+	// TODO(maisem): more protocols (sctp, gre, dccp)
+}
+
+// updateV4Checksum calculates and updates the checksum in the packet buffer
+// for a change between old and new. The checksum is updated in place.
+func updateV4Checksum(oldSum, old, new []byte) {
```
Comment (twitchyliquid64) [inline net/packet/packet.go:559]: Ditto: Add to docs that the `*Parsed` must have been decoded as an IPProto != Unknown or this is unsafe.

File: net/packet/packet.go [hunk 4]
```diff
@@ -502,3 +536,53 @@ func withIP(ap netip.AddrPort, ip netip.Addr) netip.AddrPort {
...
+	if len(old) != len(new) {
+		panic("old and new must be the same length")
+	}
+	if len(old)%2 != 0 {
+		panic("old and new must be even length")
+	}
+	/*
+	   RFC 1071 Section 1.4
```
Comment (twitchyliquid64) [inline net/packet/packet.go:567]: This terrifies me, and as a result I would like a test plz :P

File: net/tstun/wrap.go [hunk 5]
```diff
@@ -459,12 +465,153 @@ func (t *Wrapper) sendVectorOutbound(r tunVectorReadResult) {
...
+		}
+	}
+	return netip.Addr{}
+}
+
+// natV4Config is the configuration for IPv4 NAT.
+// It should be treated as immutable.
+//
+// The nil value is a valid configuration.
+type natV4Config struct {
```
Comment (twitchyliquid64) [inline net/tstun/wrap.go:518]: Should NAT really be in `tstun`? Can we have an explicitly separate layer for this?

File: net/tstun/wrap.go [hunk 6]
```diff
@@ -566,14 +713,16 @@ func (t *Wrapper) Read(buffs [][]byte, sizes []int, offset int) (int, error) {
 	defer parsedPacketPool.Put(p)
 	for _, data := range res.data {
 		p.Decode(data[res.dataOffset:])
+		if capt := t.captureHook.Load(); capt != nil {
+			capt(capture.FromLocal, time.Now(), data[res.dataOffset:])
+		}
+
+		t.snatV4(p)
```
Comment (twitchyliquid64) [inline net/tstun/wrap.go:720]: A bit further down we do:

```
n := copy(buffs[buffsPos][offset:], data[res.dataOffset:])
```

Now that `packet.Parsed` is also responsible for mutating, can we not access the underlying slice but instead do `p.Buffer()` ?

File: net/tstun/wrap.go [hunk 7]
```diff
@@ -459,12 +465,153 @@ func (t *Wrapper) sendVectorOutbound(r tunVectorReadResult) {
...
+	t.logf("nat config: %+v", cfg)
+}
+
 var (
 	magicDNSIPPort   = netip.AddrPortFrom(tsaddr.TailscaleServiceIP(), 0) // 100.100.100.100:0
 	magicDNSIPPortv6 = netip.AddrPortFrom(tsaddr.TailscaleServiceIPv6(), 0)
 )
 
 func (t *Wrapper) filterOut(p *packet.Parsed) filter.Response {
+	// packet received from the os/netstack
```
Comment (twitchyliquid64) [inline net/tstun/wrap.go:613]: Time to rename `filterOut` to something more fitting its role as an interceptor as well as a filter-er?

File: wgengine/capture/capture.go [hunk 8]
```diff
@@ -72,6 +72,10 @@ const (
 	// SynthesizedToPeer indicates the packet was generated from within tailscaled,
 	// and is being routed to a remote Wireguard peer.
 	SynthesizedToPeer Path = 3
+	// SNAT indicates the packet went through a SNAT operation.
```
Comment (twitchyliquid64) [inline wgengine/capture/capture.go:75]: I think instead of duplicating the packet in the PCAP stream we should annotate the existing `FromLocal` path with the translated before/after addresses.

To avoid a comment storm I'll hold off on capture-related comments till weve got the NAT stuff hunkydory.

File: net/tstun/wrap.go (hunk 6 above)
Comment (maisem) [reply net/tstun/wrap.go:709]: Done

File: net/packet/packet.go (hunk 4 above)
Comment (maisem) [reply net/packet/packet.go]: Agreed, working on it.

File: net/packet/packet.go [hunk 9]
```diff
@@ -502,3 +536,69 @@ func withIP(ap netip.AddrPort, ip netip.Addr) netip.AddrPort {
...
+	case ipproto.TCP:
+		updateV4Checksum(p.Transport()[16:18], o4[:], n4[:])
+	case ipproto.ICMPv4:
+		// Nothing to do.
+	}
+	// TODO(maisem): more protocols (sctp, gre, dccp)
+}
+
+// updateV4Checksum calculates and updates the checksum in the packet buffer
+// for a change between old and new. The checksum is updated in place.
```
Comment (twitchyliquid64) [inline net/packet/packet.go:560]: Was a bit confused about oldSum: Maybe mention that it must point to the 16-bit checksum bytes which are being updated, and in place?

File: net/packet/packet_test.go [hunk 10]
```diff
@@ -29,6 +30,52 @@ const (
...
+func TestHeaderChecksums(t *testing.T) {
+	// This is not a good enough test, because it doesn't
+	// check the various packet types or the many edge cases
+	// of the checksum algorithm. But it's a start.
+
+	tests := []struct {
+		name   string
+		packet []byte
+	}{
+		{
```
Comment (twitchyliquid64) [inline net/packet/packet_test.go:57]: Here's a UDP response for a DNS query to `clients4.google.com`:

```
0000   02 00 45 00 00 74 e2 85 00 00 40 11 96 b5 64 64   ..E..t....@...dd
0010   64 64 64 42 d4 33 00 35 ec 55 00 60 d9 19 ed fd   dddB.3.5.U.`....
0020   81 80 00 01 00 02 00 00 00 01 08 63 6c 69 65 6e   ...........clien
0030   74 73 34 06 67 6f 6f 67 6c 65 03 63 6f 6d 00 00   ts4.google.com..
0040   01 00 01 c0 0c 00 05 00 01 00 00 01 1e 00 0c 07   ................
0050   63 6c 69 65 6e 74 73 01 6c c0 15 c0 31 00 01 00   clients.l...1...
0060   01 00 00 01 1e 00 04 8e fa bd ce 00 00 29 02 00   .............)..
0070   00 00 00 00 00 00                                 ......

```
As hex:
```
020045000074e2850000401196b5646464646442d4330035ec550060d919edfd8180000100020000000108636c69656e74733406676f6f676c6503636f6d0000010001c00c000500010000011e000c07636c69656e7473016cc015c031000100010000011e00048efabdce0000290200000000000000
```


The first two bytes are wiresharl framing which can be trimmed.

Comment (twitchyliquid64) [inline net/packet/packet_test.go:57]: Here's a TCP SYN (connect) to mon:

```
00004500003c542940004006b1ac6442d4336461980fb19401bb0a51ce7c00000000a002fbe038f60000020404d80402080a862bccd50000000001030307
```

Again, trim the first 2 bytes.

File: net/tstun/wrap.go [hunk 11]
```diff
@@ -459,12 +465,145 @@ func (t *Wrapper) sendVectorOutbound(r tunVectorReadResult) {
...
 }
 
+// snatV4 does SNAT on p if it's an IPv4 packet and the destination
+// address requires a different source address.
+func (t *Wrapper) snatV4(p *packet.Parsed) {
+	if p.IPVersion != 4 {
+		return
+	}
+
+	nc := t.natV4Config.Load()
```
Comment (twitchyliquid64) [inline net/tstun/wrap.go:475]: `natV4Config` can be nil no? Doesn't seem to be handled?

File: net/tstun/wrap.go [hunk 12]
```diff
@@ -459,12 +465,145 @@ func (t *Wrapper) sendVectorOutbound(r tunVectorReadResult) {
...
+	}
+}
+
+// dnatV4 does destination NAT on p if it's an IPv4 packet.
+func (t *Wrapper) dnatV4(p *packet.Parsed) {
+	if p.IPVersion != 4 {
+		return
+	}
+
+	nc := t.natV4Config.Load()
```
Comment (twitchyliquid64) [inline net/tstun/wrap.go:489]: Ditto: `natV4Config` can be nil, doesn't seem to be handled?

**Review by twitchyliquid64 — COMMENTED** [review]
Woohooo almost there!!

File: net/tstun/wrap.go (hunk 11 above)
Comment (maisem) [reply net/tstun/wrap.go:475]: It's allowed to call methods on a nil NATConfig. See docs

File: net/tstun/wrap.go [hunk 13]
```diff
@@ -459,12 +465,145 @@ func (t *Wrapper) sendVectorOutbound(r tunVectorReadResult) {
...
+			return a
+		}
+	}
+	return netip.Addr{}
+}
+
+// natV4Config is the configuration for IPv4 NAT.
+// It should be treated as immutable.
+//
+// The nil value is a valid configuration.
```
Comment (maisem) [inline net/tstun/wrap.go:511]: @twitchyliquid64

File: net/packet/packet_test.go (hunk 10 above)
Comment (maisem) [reply net/packet/packet_test.go:57]: Added

Comment (maisem) [reply net/packet/packet_test.go:57]: Added

**Review by twitchyliquid64 — APPROVED** [review]

**Comment by twitchyliquid64** [comment]
Test failures tho:

```
--- FAIL: TestGetTypeHasher (0.00s)
    --- FAIL: TestGetTypeHasher/tailcfg.Node (0.00s)
        deephash_test.go:598: got "\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\tn\x88\xf1\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\tn\x88\xf1\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"; want "\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\tn\x88\xf1\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\tn\x88\xf1\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
```
